### PR TITLE
feat: add `write_engine` parameter to `read_FORMATNAME` methods to control how data is written to BigQuery

### PR DESCRIPTION
**Pull request by tswast**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes internal issue 323176126
🦕

### CODE REVIEW

**Comment by tswast** [comment]
Blocked by https://github.com/googleapis/python-bigquery/issues/1815

File: third_party/bigframes_vendored/google_cloud_bigquery/_pandas_helpers.py
Comment (tswast) [inline third_party/bigframes_vendored/google_cloud_bigquery/_pandas_helpers.py:1]: Might make sense to expose some of these methods in a public place since they are shared by the client library, pandas-gbq, and now bigframes.

A similar method is exposed publicly at

https://github.com/googleapis/python-bigquery-pandas/blob/b3b9202980a17faa9df6fc8bba7785d984452893/pandas_gbq/gbq.py#L1147

but it's been deprecated for quite some time. I think we may want to revisit that deprecation and revive it. Will need to think carefully about circular dependencies, though.

**Comment by tswast** [comment]
Test failure is a real one:

<details><summary><code>E       TypeError: Object of type bool_ is not JSON serializable</code></summary>
<pre>
=================================== FAILURES ===================================
_____________ test_read_csv_gcs_default_engine[bigquery_streaming] _____________
[gw19] linux -- Python 3.11.6 /tmpfs/src/github/python-bigquery-dataframes/.nox/system-3-11/bin/python

session = <bigframes.session.Session object at 0x7f0d0f897cd0>
scalars_dfs = (          bool_col                                          bytes_col  \
rowindex                                    ......  2038-01-19 03:14:17.999999+00:00
8            False  ...                              <NA>

[9 rows x 13 columns])
gcs_folder = 'gs://bigframes-dev-testing/bigframes_tests_system_20240229220731_1845bf/'
write_engine = 'bigquery_streaming'

    @skip_legacy_pandas
    @pytest.mark.parametrize(
        ("write_engine",),
        (
            ("default",),
            ("bigquery_inline",),
            ("bigquery_load",),
            ("bigquery_streaming",),
        ),
    )
    def test_read_csv_gcs_default_engine(session, scalars_dfs, gcs_folder, write_engine):
        scalars_df, _ = scalars_dfs
        if scalars_df.index.name is not None:
            path = gcs_folder + "test_read_csv_gcs_default_engine_w_index*.csv"
        else:
            path = gcs_folder + "test_read_csv_gcs_default_engine_wo_index*.csv"
        read_path = path.replace("*", FIRST_FILE)
        scalars_df.to_csv(path, index=False)
        dtype = scalars_df.dtypes.to_dict()
        dtype.pop("geography_col")
>       df = session.read_csv(
            read_path,
            # Convert default pandas dtypes to match BigQuery DataFrames dtypes.
            dtype=dtype,
            write_engine=write_engine,
        )

[tests/system/small/test_session.py:435](https://cs.corp.google.com/piper///depot/google3/tests/system/small/test_session.py?l=435): 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[bigframes/session/__init__.py:1162](https://cs.corp.google.com/piper///depot/google3/bigframes/session/__init__.py?l=1162): in read_csv
    return self._read_pandas(
[bigframes/session/__init__.py:933](https://cs.corp.google.com/piper///depot/google3/bigframes/session/__init__.py?l=933): in _read_pandas
    return self._read_pandas_bigquery_table(
[bigframes/session/__init__.py:988](https://cs.corp.google.com/piper///depot/google3/bigframes/session/__init__.py?l=988): in _read_pandas_bigquery_table
    table_expression = bigframes_io.pandas_to_bigquery_streaming(
[bigframes/session/_io/bigquery.py:294](https://cs.corp.google.com/piper///depot/google3/bigframes/session/_io/bigquery.py?l=294): in pandas_to_bigquery_streaming
    for errors in bqclient.insert_rows_from_dataframe(
[.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py:3662](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py?l=3662): in insert_rows_from_dataframe
    result = self.insert_rows(table, rows_chunk, selected_fields, **kwargs)
[.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py:3605](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py?l=3605): in insert_rows
    return self.insert_rows_json(table, json_rows, **kwargs)
[.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py:3801](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py?l=3801): in insert_rows_json
    response = self._call_api(
[.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py:827](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/cloud/bigquery/client.py?l=827): in _call_api
    return call()
[.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py:293](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py?l=293): in retry_wrapped_func
    return retry_target(
[.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py:153](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py?l=153): in retry_target
    _retry_error_helper(
[.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_base.py:212](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_base.py?l=212): in _retry_error_helper
    raise final_exc from source_exc
[.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py:144](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py?l=144): in retry_target
    result = target()
[.nox/system-3-11/lib/python3.11/site-packages/google/cloud/_http/__init__.py:479](https://cs.corp.google.com/piper///depot/google3/.nox/system-3-11/lib/python3.11/site-packages/google/cloud/_http/__init__.py?l=479): in api_request
    data = json.dumps(data)
/[usr/local/lib/python3.11/json/__init__.py:231](https://cs.corp.google.com/piper///depot/google3/usr/local/lib/python3.11/json/__init__.py?l=231): in dumps
    return _default_encoder.encode(obj)
/[usr/local/lib/python3.11/json/encoder.py:200](https://cs.corp.google.com/piper///depot/google3/usr/local/lib/python3.11/json/encoder.py?l=200): in encode
    chunks = self.iterencode(o, _one_shot=True)
/[usr/local/lib/python3.11/json/encoder.py:258](https://cs.corp.google.com/piper///depot/google3/usr/local/lib/python3.11/json/encoder.py?l=258): in iterencode
    return _iterencode(o, 0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <json.encoder.JSONEncoder object at 0x7f0d3d5e8b90>, o = True

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).
    
        For example, to support arbitrary iterators, you could
        implement default like this::
    
            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)
    
        """
>       raise TypeError(f'Object of type {o.__class__.__name__} '
                        f'is not JSON serializable')
E       TypeError: Object of type bool_ is not JSON serializable

/[usr/local/lib/python3.11/json/encoder.py:180](https://cs.corp.google.com/piper///depot/google3/usr/local/lib/python3.11/json/encoder.py?l=180): TypeError
----------------------------- Captured stdout call -----------------------------
Query job 2b56fdd1-5dcc-4afd-8521-69b517d5afae is DONE.1.1 kB processed.  
https://console.cloud.google.com/bigquery?project=bigframes-dev&j=bq:US:2b56fdd1-5dcc-4afd-8521-69b517d5afae&page=queryresults
Query job 7d127271-a535-48d0-b092-df329771fb89 is RUNNING. 
https://console.cloud.google.com/bigquery?project=bigframes-dev&j=bq:US:7d127271-a535-48d0-b092-df329771fb89&page=queryresults
Query job 7d127271-a535-48d0-b092-df329771fb89 is DONE.0 Bytes processed.  
https://console.cloud.google.com/bigquery?project=bigframes-dev&j=bq:US:7d127271-a535-48d0-b092-df329771fb89&page=queryresults
=============================== warnings summary ===============================
</pre>
</details>

This will likely require a fix upstream in google-cloud-bigquery, but in the meantime I can make sure to use a vendored version of `insert_rows_from_dataframe` that can serialize a numpy `bool_` value, similar to how there's a special case for `NaN` already.

File: third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py
Comment (shobsi) [inline third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py:48]: This file doesn't seem to be added yet

File: third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py
Comment (shobsi) [inline third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py:94]: Why not add in the first declaration itself?

File: third_party/bigframes_vendored/google_cloud_bigquery/tests/unit/test_pyarrow_hlpers.py
Comment (shobsi) [inline third_party/bigframes_vendored/google_cloud_bigquery/tests/unit/test_pyarrow_hlpers.py:30]: shouldn't all types be unit tested?

File: bigframes/session/validation.py
Comment (shobsi) [inline bigframes/session/validation.py:18]: Should we make it more helpful by suggesting what are valid combinations?

File: tests/unit/session/test_session.py
Comment (shobsi) [inline tests/unit/session/test_session.py:31]: Add (None, "bigquery_external_table") too?

File: third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py
Comment (tswast) [reply third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py:48]: This is copied directly from upstream https://github.com/googleapis/python-bigquery repo.

It's referring to this file: https://github.com/googleapis/python-bigquery-storage/blob/main/tests/unit/test_reader_v1_arrow.py#L44

Note: I have proposed go/pandas-gbq-and-bigframes-redundancy so that we won't have to keep this copy around. Will chat with the BigQuery SDK folks more about it on Thursday.

File: third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py
Comment (tswast) [reply third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py:94]: This is copied from https://github.com/googleapis/python-bigquery repo with the change that I removed the conditional if the necessary version of pyarrow is not installed. I'd like to keep the file as close to upstream as possible.

File: third_party/bigframes_vendored/google_cloud_bigquery/tests/unit/test_pyarrow_hlpers.py
Comment (tswast) [reply third_party/bigframes_vendored/google_cloud_bigquery/tests/unit/test_pyarrow_hlpers.py:30]: Maybe? That would mostly involve duplicating the source dictionary, though. Again, copied from upstream. We can make the patch there, potentially as part of go/pandas-gbq-and-bigframes-redundancy.

Since there are open questions,  I don't believe this PR needs to block GA. We can wait for that design to be resolved before merging this.

File: third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py
Comment (shobsi) [reply third_party/bigframes_vendored/google_cloud_bigquery/_pyarrow_helpers.py:48]: Got it, thanks

File: tests/unit/session/test_session.py
Comment (tswast) [reply tests/unit/session/test_session.py:31]: Haven't implemented that one yet.

**Comment by tswast** [comment]
> Marking as do not merge for now. Thanks for your feedback so far. I will wait until we implement go/pandas-gbq-and-bigframes-redundancy before merging this (likely mid-April).

I've mailed https://github.com/googleapis/python-bigquery-pandas/pull/814 as a first step of this project. I plan to follow-up that PR with one that copies the `bigquery_streaming` from this PR to pandas-gbq so that the pandas -> BigQuery logic can be consolidated to the pandas-gbq package.

**Review by OKEAMAH — APPROVED** [review]

**Comment by tswast** [comment]
@shobsi This is ready for another look. Not quite merge-ready yet, as I need to update pandas-gbq's schema detection to account for `ArrowDtype(pa.timestamp(...))` (https://github.com/googleapis/python-bigquery-pandas/issues/824).

File: third_party/bigframes_vendored/google_cloud_bigquery/_pandas_helpers.py
Comment (tswast) [reply third_party/bigframes_vendored/google_cloud_bigquery/_pandas_helpers.py:1]: Done in https://github.com/googleapis/python-bigquery-pandas/pull/814

File: bigframes/session/__init__.py
Comment (tswast) [inline bigframes/session/__init__.py:742]:
```suggestion
                "read_pandas",
                write_engine=write_engine,
```

File: bigframes/session/validation.py
Comment (tswast) [reply bigframes/session/validation.py]: I worry that will be hard to maintain. Perhaps we can move some of the combinations as constants, too?

**Comment by tswast** [comment]
Looks like there are a few more tests for me to cleanup. These do look related to my change:

```
FAILED tests/system/small/test_session.py::test_read_pandas_w_unsupported_mixed_dtype
FAILED tests/system/small/test_series.py::test_series_explode_null[empty_array]
FAILED tests/system/small/test_series.py::test_series_explode_null[null_and_empty_array]
```

File: tests/unit/session/test_session.py
Comment (shobsi) [reply tests/unit/session/test_session.py:168]: Looks like we do throw the same exception from `validate_engine_compatibility`, which means this unit test extends to all these combinations:

({None, "c", "python", "pyarrow", "python-fwf"}, "bigquery_external_table")

File: third_party/bigframes_vendored/constants.py
Comment (shobsi) [inline third_party/bigframes_vendored/constants.py:43]: [nit] can we do something like this to save defining twice?

```python
VALID_WRITE_ENGINES = list(typing.get_args(WriteEngineType))
```

**Review by tswast — COMMENTED** [review]
Updated. Thanks for the review!

Looks like I still need a fix in pandas-gbq for empty arrays.

**Comment by tswast** [comment]
https://github.com/googleapis/python-bigquery-pandas/pull/838 to fix the remaining system test. Tested locally and it works.

File: bigframes/session/validation.py
Comment (shobsi) [reply bigframes/session/validation.py]: resolving the outdated comment

File: third_party/bigframes_vendored/google_cloud_bigquery/tests/unit/test_pyarrow_hlpers.py
Comment (shobsi) [reply third_party/bigframes_vendored/google_cloud_bigquery/tests/unit/test_pyarrow_hlpers.py:1]: resolving outdated comment

**Review by shobsi — APPROVED** [review]
LGTM to merge after the pandas-gbq fix and update

File: bigframes/dtypes.py
Comment (tswast) [inline bigframes/dtypes.py:413]: @GarrettWu Would this break the JSON support you're adding? It was needed for some failing CSV tests.

**Pull request merged**
